crates/sui-graphql-rpc/src/extensions/query_limits_checker.rs

```diff
@@ -124,49 +136,31 @@ impl Extension for QueryLimitsChecker {
             ));
         }
 
-        // Use BFS to analyze the query and
-        // count the number of nodes and the depth of the query
-
-        // Queue to store the nodes at each level
-        let mut que = VecDeque::new();
-        // Number of nodes in the query
-        let mut num_nodes: u32 = 0;
-        // Depth of the query
-        let mut depth: u32 = 0;
-        // Number of nodes at each level
-        let mut level_len;
-
-        for (_name, oper) in doc.operations.iter() {
-            for sel in oper.node.selection_set.node.items.iter() {
-                que.push_back(sel);
-                num_nodes += 1;
-                self.check_limits(cfg, num_nodes, depth, Some(sel.pos))?;
+        // Only allow one operation
+        match doc.operations.iter().count() {
+            0 => {
+                return Err(ServerError::new(
+                    "One operation is required".to_string(),
+                    None,
+                ));
             }
-        }
-        // Track the number of nodes at first level if any
-        level_len = que.len();
-
-        while !que.is_empty() {
-            // Signifies the start of a new level
-            depth += 1;
-            self.check_limits(cfg, num_nodes, depth, None)?;
-            while level_len > 0 {
-                // Ok to unwrap since we checked for empty queue
-                // and level_len > 0
-                let sel = que.pop_front().unwrap();
-                // TODO: check for fragments, variables, etc
-                if let Field(f) = &sel.node {
-                    // TODO: check for directives, variables, etc
-                    for sel in f.node.selection_set.node.items.iter() {
-                        que.push_back(sel);
-                        num_nodes += 1;
-                        self.check_limits(cfg, num_nodes, depth, Some(sel.pos))?;
-                    }
-                }
-                level_len -= 1;
+            1 => {}
+            _ => {
+                return Err(ServerError::new(
+                    "Query has too many operations. The maximum allowed is 1".to_string(),
+                    None,
+                ));
             }
-            level_len = que.len();
         }
+
+        // TODO: Limit the complexity of fragments early on
+
+        // Okay to unwrap since we checked for 1 operation
+        let (_name, oper) = doc.operations.iter().next().unwrap();
```

Comment:
Avoid the unwrap with:

```rust
let mut operations = doc.operations.iter()
let Some((_name, oper)) = operations.next() else {
    return Err(graphql_error(code::BAD_REQUEST, "Query must have one operation");
};

if operations.next().is_some() {
    return Err(graphql_error("Query has too many operations. The maximum allowed is 1"));
}
```

Although I'm a little confused why the query limit checker needs to impose these limits at all.  If there are no operations, can we just treat it as having zero nodes, zero depth, and if there are multiple can't we combine the depths and nodes from all of them, if that's allowed at all?

Reply:
This is the only outstanding comment for me -- do we need to impose the limit on number of operations in a query at this stage?  It seems like even if this is something we wanted to do, if we did it here we get a worse error message than we might otherwise get.

Reply:
I've removed the fragment limit and combined the costs of all fragments.